### ISSUE TITLE
Create alias for 'when' so that no escaping is required in Kotlin

### ISSUE THREAD

**Issue by jasperblues** [rest-assured/rest-assured#609]
Kotlin has a 'when' control flow in the language, so when must be escaped with `when`

This could be fixed by using an alias that calls when?

**Comment by johanhaleby** [comment]
I recently created a scala support module that added a "Then" method alias to "then". Does Kotlin support implicit functions/methods? In that case we might be able to do something similar.

**Comment by jasperblues** [comment]
Yes it does, though it calls them class extensions, like Swift, Groovy? and others.

**Comment by johanhaleby** [comment]
Cool. Would it be nice to have a module similar to the [scala module](https://github.com/jayway/rest-assured/wiki/Usage#scala-support-module) that does this? Or is it so simple that perhaps it's better to just document it in the documentation?

**Comment by jasperblues** [comment]
It is super-simple, and documentation would do it. But the reason for the suggestion was that I thought it might be nice to agree on standard names, to make it easy to go from one code-base to another.

**Comment by johanhaleby** [comment]
I understand, but I suppose the problem is that different languages have different reserved keywords and all of a sudden we find that in language X `now` is reserved keyword.

**Comment by jasperblues** [comment]
:+1: yeah, probably. A note in the Readme couldn't hurt? Or it is obvious what to do?

**Comment by johanhaleby** [comment]
I think it would fit nicely in the usage guide. I've written a small blog about it [here](http://blog.haleby.se/2015/11/01/rest-assured-with-kotlin/). What do you think? Is trait a good way to add the alias or is there a better way in Kotlin? I'll add it to the docs when/if you approve my approach in the blog :)

**Comment by johanhaleby** [comment]
I've not documented this (see frontpage) so I'm going to close the issue

**Comment by jasperblues** [comment]
not or have :P

**Comment by johanhaleby** [comment]
have, sorry :) I always seem to type too fast for my own good.

**Comment by jasperblues** [comment]
don't we all ;) Thanks again for this library - very useful.

**Comment by fmobus** [comment]
Just for future reference for any kotlin devs out there that might want to make a DSL out of rest-assured. You can achieve something like

````kotlin
Given {
    port(7000)
    header("Authorization", "bla")
    body(myObj, mapper)
} When {
    put("/the/path")
} Then {
    statusCode(200)
    body("message", equalTo("Hello World"))
}
````

With just a handful of lines of kotlin magic (quick and dirty version, use at your own risk):
````kotlin
fun Given(block: RequestSpecification.() -> RequestSpecification) = given().run(block)
infix fun RequestSpecification.When(block: RequestSpecification.() -> Response) = this.`when`().run(block)
infix fun Response.Then(block: ValidatableResponse.() -> Unit) = this.then().run(block)
````

**Comment by johanhaleby** [comment]
@fmobus That's actually pretty cool, and useful. When using REST Assured in Kotlin I usually hack something like this together in the project I'm working. So maybe we should try to standardize it a bit more.

Wouldn't this also solve the annoying indentation problem? I.e. when you try to format rest assured tests written in Java they can look quite ugly. I guess this would work? If so it's even better :)

**Comment by johanhaleby** [comment]
Another awesome benefit is that it *should* be possible to report all errors in the same go due to:

```kotlin
Then {
    statusCode(200)
    body("message", equalTo("Hello World"))
}
```

**Comment by johanhaleby** [comment]
I've now added initial support for this. Please try version `4.0.1-SNAPSHOT` after having added the following maven repo:

```xml
<repositories>
        <repository>
            <id>sonatype</id>
            <url>https://oss.sonatype.org/content/repositories/snapshots/</url>
            <snapshots />
        </repository>
</repositories>
```

Then add the new `kotlin-extensions` module:

```
<dependency>
    <groupId>io.rest-assured</groupId>
    <artifactId>kotlin-extensions</artifactId>
    <version>${rest-assured.version}</version>
    <scope>test</scope>
</dependency>
```
and then import `Given` from the `io.restassured.module.kotlin.extensions` package. You can then use it like this:

```kotlin
val message: String =
Given {
  port(7000)
  header("Header", "Header")
  body("hello")
  filter { _, _, _ ->
      ResponseBuilder().setStatusCode(200).setContentType(JSON).setBody("""{ "message" : "Hello World"}""").build()
  }
} When {
  put("/the/path")
} Extract {
  path("message")
}
```

**Comment by johanhaleby** [comment]
I've now also implemented so that all failures are reported in the same go when using `Then`. I.e. if 

```kotlin
Then {
    statusCode(200)
    body("message", equalTo("Hello World"))
}
```

fails then both the status code and the body failure are reported. This is NOT the case with the Java API.
